client/network/src/request_responses.rs

```diff
@@ -345,148 +351,209 @@ impl NetworkBehaviour for RequestResponsesBehaviour {
+						return Poll::Ready(NetworkBehaviourAction::ReportObservedAddr {
+							address, score,
+						})
+					}
+				};
 
-					match ev {
-						// Received a request from a remote.
-						RequestResponseEvent::Message {
-							peer,
-							message: RequestResponseMessage::Request { request, channel, .. },
-						} => {
-							let (tx, rx) = oneshot::channel();
-
-							// Submit the request to the "response builder" passed by the user at
-							// initialization.
-							if let Some(resp_builder) = resp_builder {
-								// If the response builder is too busy, silently drop `tx`.
-								// This will be reported as a `Busy` error.
-								let _ = resp_builder.try_send(IncomingRequest {
-									peer: peer.clone(),
-									payload: request,
-									pending_response: tx,
-								});
-							}
+				match ev {
+					// Received a request from a remote.
+					RequestResponseEvent::Message {
+						peer,
+						message: RequestResponseMessage::Request { request_id, request, channel, .. },
+					} => {
+						self.pending_responses_arrival_time.insert(
+							request_id.clone(),
+							Instant::now(),
+						);
+
+						let (tx, rx) = oneshot::channel();
+
+						// Submit the request to the "response builder" passed by the user at
+						// initialization.
+						if let Some(resp_builder) = resp_builder {
+							// If the response builder is too busy, silently drop `tx`.
+							// This will be reported as a `Busy` error.
+							let _ = resp_builder.try_send(IncomingRequest {
+								peer: peer.clone(),
+								payload: request,
+								pending_response: tx,
+							});
+						}
 
-							let protocol = protocol.clone();
-							self.pending_responses.push(Box::pin(async move {
-								// The `tx` created above can be dropped if we are not capable of
-								// processing this request, which is reflected as a "Busy" error.
-								if let Ok(response) = rx.await {
-									RequestProcessingOutcome::Response {
-										protocol, inner_channel: channel, response
-									}
-								} else {
-									RequestProcessingOutcome::Busy { peer, protocol }
+						let protocol = protocol.clone();
+						self.pending_responses.push(Box::pin(async move {
+							// The `tx` created above can be dropped if we are not capable of
+							// processing this request, which is reflected as a "Busy" error.
+							if let Ok(response) = rx.await {
+								RequestProcessingOutcome::Response {
+									request_id, peer, protocol, inner_channel: channel, response
 								}
-							}));
+							} else {
+								RequestProcessingOutcome::Busy { peer, protocol }
+							}
+						}));
 
-							// This `continue` makes sure that `pending_responses` gets polled
-							// after we have added the new element.
-							continue 'poll_all;
-						}
+						// This `wake_by_ref` makes sure that `pending_responses` gets polled
+						// after we have added the new element.
+						cx.waker().wake_by_ref();
```

Comment:
Want to draw attention to this change. As far as I can tell this is good enough to force the `poll` function to be called again.

Reply:
According to the docs, that is indeed correct: https://docs.rs/futures/0.3.8/futures/task/struct.Waker.html#method.wake_by_ref

Reply:
I initially made this change for the cleanup step of `self.pending_responses_arrival_time`. Given that the cleanup was an additional step running after all other steps I thought removing the extra loop would reduce complexity. This was before https://github.com/paritytech/substrate/pull/7696/commits/2b8b0ff61983ca0143d78a4b2dfe1eb31ab24ec5 using an `LruCache` instead, thus making the cleanup step obsolete.

Today with https://github.com/paritytech/substrate/pull/7696/commits/2b8b0ff61983ca0143d78a4b2dfe1eb31ab24ec5 I still think that this change simplifies the control flow and makes the code more readable due to the reduced indentation levels. This is highly subjective. I am happy to bring back the loop.

What do you think @romanb?

Reply:
It just seems to me that as per the `FuturesUnordered::push()` contract it is desirable to poll it again immediately after pushing a new item. Scheduling another task wakeup to do so seems like going through an unnecessary delay / jumping through an extra hoop. Hence I wondered if there may be the reason for this delay actually being desirable. It is always nice to remove a level of indentation / looping, but this is certainly a code transformation that changes the runtime behaviour. Don't feel compelled to revert the change though if you think the extra task wakeup is beneficial somehow.

Reply:
> It is always nice to remove a level of indentation / looping, but this is certainly a code transformation that changes the runtime behaviour.

Good point. Especially as our future hierarchies tend to be quite large. This is reverted with 30c1aeada65834416a99be214858f3216551a688.

Sorry for making the review process hard with this noise change.